go/processor/geth/processor.go

```
@@ -0,0 +1,397 @@
+	}
+	return config
+}
+
+func transactionToMessage(transaction tosca.Transaction, baseFee tosca.Value) *core.Message {
+	accessList := types.AccessList{}
+	for _, tuple := range transaction.AccessList {
+		storageKeys := make([]common.Hash, len(tuple.Keys))
+		for i, key := range tuple.Keys {
+			storageKeys[i] = common.Hash(key)
+		}
+		accessList = append(accessList, types.AccessTuple{
+			Address:     common.Address(tuple.Address),
+			StorageKeys: storageKeys,
+		})
+	}
+
+	return &core.Message{
+		From:     common.Address(transaction.Sender),
+		To:       (*common.Address)(transaction.Recipient),
+		Nonce:    transaction.Nonce,
+		Value:    transaction.Value.ToBig(),
+		GasLimit: uint64(transaction.GasLimit),
+		GasPrice: transaction.GasPrice.ToBig(),
+		// gas price computation and enforcement of cap is currently performed outside of processor
+		GasFeeCap:         big.NewInt(0).Add(baseFee.ToBig(), big.NewInt(1)),
+		GasTipCap:         big.NewInt(0),
+		Data:              transaction.Input,
+		AccessList:        accessList,
+		BlobGasFeeCap:     big.NewInt(0),
+		BlobHashes:        nil,
+		SkipAccountChecks: false,
+	}
+}
+
+// stateDB is a wrapper around the tosca.TransactionContext to implement the vm.StateDB interface.
+type stateDB struct {
+	context         tosca.TransactionContext
+	refund          uint64
+	createdContract common.Address
+	refundBackups   map[tosca.Snapshot]uint64
+	beneficiary     common.Address
+}
+
+// vm.StateDB interface implementation
+
+func (s *stateDB) CreateAccount(common.Address) {
+	// not implemented
+}
+
+func (s *stateDB) CreateContract(address common.Address) {
+	s.createdContract = address
+}
+
+func (s *stateDB) SubBalance(address common.Address, value *uint256.Int, tracing tracing.BalanceChangeReason) {
+	toscaAddress := tosca.Address(address)
+	balance := s.context.GetBalance(toscaAddress)
+	s.context.SetBalance(toscaAddress, tosca.Sub(balance, tosca.ValueFromUint256(value)))
```

Should there be a check that the available balance is higher than the balance to be subtracted?

---

This function is only called after a check for sufficient balance has been performed. I can not add an error due to the interface and I don't want the function to panic and kill the whole process.

---

The `stateDB` interface does not provide the typical error handling support where each function would return an error. This would make the handling of errors very complicated while running an EVM contract.

Instead, the idea is that errors are collected internally and operations are skipped if something fails. At the end of a transaction it is checked whether any errors have been encountered, and if so, the transaction is reverted. However, this error-check function is not part of the `StateDB` interface provided to the processor.

If we want the same error handling guarantees, we would have to do the same here. Have an error field in the `stateDB` and collect errors that are checked at the end of the processor run. Or, since it is a reference implementation that should not run in production and only needs to fit a single use case, a panic might be the way to go.

Ignoring it may be an option as well, but it pushes the responsibility to check that the balance is high enough to the caller.

---

There should not be a call to this function outside of the `transfer` function in the `blockContext` and the `state contract`. Both perform a balance check before. I would ignore the error here, as the `context` implementation used in production also wont panic or set an error, it is the processors responsibility to ensure the balance is sufficient.